### PR TITLE
[MOB-47] Mock subscriptions payment flow

### PR DESCRIPTION
**Pull request by carvalho2707**
**What does this PR do?**

   This PR implements subscriptions for the payments methods flow

**Database changed?**

No

**Where should the reviewer start?**

- [ ] IabActivity.kt

**How should this be manually tested?**

  Flow on how to test this or QA Tickets related to this use-case: [MOB-47](https://aptoide.atlassian.net/browse/MOB-47)

**What are the relevant tickets?**

  Tickets related to this pull-request: [MOB-47](https://aptoide.atlassian.net/browse/APPC-MOB-47)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt [hunk 1]
```diff
@@ -373,6 +385,11 @@ class AdyenPaymentFragment : DaggerFragment(), AdyenPaymentView {
     bonus_msg?.visibility = View.VISIBLE
     bonus_msg_pre_selected?.visibility = View.VISIBLE
     bonus_value.text = getString(R.string.gamification_purchase_header_part_2, bonus)
+
+    frequency?.let {
+      bonus_msg?.text = "You will receive this bonus for each payment"
+      bonus_msg_pre_selected?.text = "You will receive this bonus for each payment"
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt]: missing final string

File: app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt [hunk 2]
```diff
@@ -610,4 +633,12 @@ class AdyenPaymentFragment : DaggerFragment(), AdyenPaymentView {
       throw IllegalArgumentException("pre selected data not found")
     }
   }
+
+  private val frequency: String? by lazy {
+    if (arguments!!.containsKey(FREQUENCY)) {
+      arguments!!.getString(FREQUENCY)
+    } else {
+      null
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt:641]: You can initialize it as null instead

File: app/src/main/java/com/asfoundation/wallet/ui/iab/IabActivity.kt [hunk 3]
```diff
@@ -174,11 +169,13 @@ class IabActivity : BaseActivity(), IabView, UriNavigator {
...
     val isDonation = TransactionData.TransactionType.DONATION.name
         .equals(transaction?.type, ignoreCase = true)
+    val isSubscription = TransactionData.TransactionType.SUBS.name
+        .equals(transaction?.type, ignoreCase = true)
     supportFragmentManager.beginTransaction()
         .replace(R.id.fragment_container, PaymentMethodsFragment.newInstance(transaction,
             intent.extras!!
                 .getString(PRODUCT_NAME), isBds, isDonation, developerPayload, uri,
-            intent.dataString))
+            intent.dataString, isSubscription, "Month"))
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/IabActivity.kt]: Do you want Month? As it can be weekly, yearly...

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java [hunk 4]
```diff
@@ -467,10 +477,10 @@ private CharSequence getApplicationName(String appPackage)
         transaction.getCallbackUrl(), transaction.getOrderReference(), transaction.getPayload());
   }
 
-  @Override public void setBonus(@NotNull BigDecimal bonus, @NotNull String currency) {
+  @Override public void setBonusPurchase(@NotNull BigDecimal bonus, @NotNull String currency) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java]: setPurchaseBonus instead

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java [hunk 5]
```diff
@@ -467,10 +477,10 @@ private CharSequence getApplicationName(String appPackage)
...
     BigDecimal scaledBonus = bonus.stripTrailingZeros()
         .setScale(2, BigDecimal.ROUND_DOWN);
-    if (scaledBonus.compareTo(new BigDecimal(0.01)) < 0) {
+    if (scaledBonus.compareTo(new BigDecimal("0.01")) < 0) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java:485]: Why? (Curiosity)

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -237,6 +237,7 @@
   <string name="transaction_type_poa_offchain">@string/transaction_type_poa</string>
   <string name="transaction_type_standard">Transfer</string>
   <string name="transaction_type_bonus">Bonus</string>
+  <string name="transaction_type_subscription">Subscription</string>
```
Comment (JDSM01) [inline app/src/main/res/values/strings.xml]: I think this is not being used

**Review by JDSM01 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt (hunk 2 above)
Comment (JoaoRaimundo) [reply app/src/main/java/com/asfoundation/wallet/billing/adyen/AdyenPaymentFragment.kt:641]: Initialising it with a value, defeats the purpose of being lazy

File: app/src/main/java/com/asfoundation/wallet/ui/iab/MergedAppcoinsPresenter.kt [hunk 7]
```diff
@@ -85,7 +86,7 @@ class MergedAppcoinsPresenter(private val view: MergedAppcoinsView,
 
   private fun handlePaymentSelectionChange() {
     disposables.add(view.getPaymentSelection()
-        .doOnNext { handleSelection(it) }
+        .doOnNext { handleSelection(it, isSubscription) }
```
Comment (JoaoRaimundo) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/MergedAppcoinsPresenter.kt]: Why do we pass the `isSubscription` as argument to the method when the variable is local and does not change?

File: app/src/main/java/com/asfoundation/wallet/ui/iab/MergedAppcoinsView.kt [hunk 8]
```diff
@@ -8,7 +8,7 @@ interface MergedAppcoinsView {
   fun showError(@StringRes errorMessage: Int)
   fun getPaymentSelection(): Observable<String>
   fun hideBonus()
-  fun showBonus()
+  fun showBonus(isSubscription: Boolean)
```
Comment (JoaoRaimundo) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/MergedAppcoinsView.kt]: Consider changing the name of the argument, although it is related with `isSubscription` for this design, in the UI it maybe reused for other purpose.

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java [hunk 9]
```diff
@@ -508,6 +534,16 @@ private CharSequence getApplicationName(String appPackage)
   @Override public void showBonus() {
     bonusView.setVisibility(View.VISIBLE);
     bonusMsg.setVisibility(View.VISIBLE);
+    bonusMsg.setText(R.string.gamification_purchase_body);
+    if (noBonusMsg != null) {
+      noBonusMsg.setVisibility(View.INVISIBLE);
+    }
+  }
+
+  @Override public void showBonusSubscription() {
```
Comment (JoaoRaimundo) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java]: This method can be avoided by using the `showBonus ` and passing as argument the string it should be used on the text view

File: app/src/main/res/layout/dialog_buy_buttons.xml [hunk 10]
```diff
@@ -1,12 +1,14 @@
...
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
     android:clipChildren="false"
     android:clipToPadding="false"
     android:gravity="end"
     android:orientation="horizontal"
     android:padding="@dimen/normal_margin"
+    tools:showIn="@layout/dialog_credit_card_authorization"
```
Comment (JoaoRaimundo) [inline app/src/main/res/layout/dialog_buy_buttons.xml]: Does this layout exists?

File: app/src/main/res/layout/dialog_buy_buttons.xml [hunk 11]
```diff
@@ -25,12 +26,10 @@
 
   <Button
       android:id="@+id/buy_button"
-      style="@style/ButtonStyle.Alternative"
+      style="@style/ButtonStyle"
```
Comment (JoaoRaimundo) [inline app/src/main/res/layout/dialog_buy_buttons.xml]: The difference between this two layout is that the disable state for the alternative is grey and for the standard is invisible. Is this change intentional?

File: app/src/main/res/layout/dialog_buy_buttons_payment_methods.xml [hunk 12]
```diff
@@ -18,7 +18,6 @@
       android:layout_height="@dimen/button_height"
       android:layout_gravity="center_vertical"
       android:layout_marginEnd="@dimen/big_margin"
-      android:maxWidth="126dp"
       android:minWidth="80dp"
```
Comment (JoaoRaimundo) [inline app/src/main/res/layout/dialog_buy_buttons_payment_methods.xml:21]: Why did we removed the maxSize? If this is removed, for buttons with large text, the button may grow indefinitely.

File: app/src/main/res/layout/dialog_buy_buttons_payment_methods.xml [hunk 13]
```diff
@@ -31,7 +30,6 @@
       android:layout_width="wrap_content"
       android:layout_height="35dp"
       android:layout_gravity="center_vertical"
-      android:maxWidth="126dp"
```
Comment (JoaoRaimundo) [inline app/src/main/res/layout/dialog_buy_buttons_payment_methods.xml:34]: The same comment as above.

File: app/src/main/res/values/strings.xml (hunk 6 above)
Comment (JoaoRaimundo) [reply app/src/main/res/values/strings.xml]: Be careful when removing strings (or adding) to avoid conflicts with the strings set by the copy team.

File: billing/src/main/java/com/appcoins/wallet/billing/repository/entity/TransactionData.kt [hunk 14]
```diff
@@ -28,6 +28,7 @@ class TransactionData @JvmOverloads constructor(@SerializedName("type")
   enum class TransactionType {
     DONATION,
     INAPP,
+    SUBS,
```
Comment (JoaoRaimundo) [inline billing/src/main/java/com/appcoins/wallet/billing/repository/entity/TransactionData.kt]: Consider adding a comment here for the time when the subscriptions are supported by the backend and make sure that this will be a TransactionType

**Review by JoaoRaimundo — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java (hunk 5 above)
Comment (carvalho2707) [reply app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java:485]: https://stackoverflow.com/a/7186298/2002274

File: app/src/main/res/values/strings.xml (hunk 6 above)
Comment (carvalho2707) [reply app/src/main/res/values/strings.xml]: This is being used in **TransactionDetailActivity** but it should be removed when the new screens are ready

**Review by JoaoRaimundo — APPROVED** [review]
